klib/syslog.c

```diff
@@ -0,0 +1,338 @@
+#include <unix_internal.h>
+#include <drivers/console.h>
+#include <filesystem.h>
+#include <lwip.h>
+#include <mktime.h>
+
+#define __STRING(x)     #x
+#define __XSTRING(x)    __STRING(x)
+
+#define SYSLOG_BUF_LEN  (8 * KB)
+
+#define SYSLOG_FLUSH_INTERVAL   seconds(1)
+
+#define SYSLOG_FILE_MAXSIZE (8 * MB)
+
+/* facility: 1 (user-level messages)
+ * severity: 6 (informational messages)
+ * priority = facility * 8 + severity
+ */
+#define SYSLOG_PRIORITY 14
+
+#define SYSLOG_VERSION  1
+
+#define SYSLOG_UDP_PORT 514
+
+declare_closure_struct(0, 1, void, syslog_timer_func,
+    u64, overruns);
+
+#define syslog_lock()   u64 _irqflags = spin_lock_irq(&syslog.lock)
+#define syslog_unlock() spin_unlock_irq(&syslog.lock, _irqflags)
+
+static struct {
+    struct console_driver driver;
+    heap h;
+    buffer output_file;
+    sg_io fs_write;
+    u64 file_offset;
+    sg_list file_sg;
+    sg_buf file_sgb;
+    timer file_flush_timer;
+    closure_struct(syslog_timer_func, file_flush);
+    buffer program;
+    char *server;
+    boolean dns_in_progress;
+    ip_addr_t server_ip;
+    struct udp_pcb *udp_pcb;
+    char local_ip[40];
+    bytes hdr_len;
+    struct spinlock lock;
+} syslog;
+
+static struct {
+    fsfile (*fsfile_open_or_create)(buffer file_path);
+    sg_io (*fsfile_get_writer)(fsfile f);
+    sg_list (*allocate_sg_list)(void);
+    sg_buf (*sg_list_tail_add)(sg_list sg, word length);
+    void (*deallocate_sg_list)(sg_list sg);
+    void (*runtime_memcpy)(void *a, const void *b, bytes len);
+    timer (*register_timer)(clock_id id, timestamp val, boolean absolute, timestamp interval,
+            timer_handler n);
+    void (*timm_dealloc)(tuple t);
+    sysreturn (*fs_rename)(buffer oldpath, buffer newpath);
+    err_t (*dns_gethostbyname)(const char *hostname, ip_addr_t *addr,
+            dns_found_callback found, void *callback_arg);
+    struct netif *(*netif_get_default)(void);
+    char *(*ipaddr_ntoa_r)(const ip_addr_t *addr, char *buf, int buflen);
+    struct pbuf *(*pbuf_alloc)(pbuf_layer layer, u16_t length, pbuf_type type);
+    u8 (*pbuf_free)(struct pbuf *p);
+    timestamp (*now)(clock_id id);
+    struct tm *(*gmtime_r)(u64 *timep, struct tm *result);
+    int (*rsnprintf)(char *str, u64 size, const char *fmt, ...);
+    err_t(*udp_sendto)(struct udp_pcb *pcb, struct pbuf *p,
+            const ip_addr_t *dst_ip, u16_t dst_port);
+} kfuncs;
+
+closure_function(3, 1, void, syslog_file_write_complete,
+                 void *, buf, u64, len, sg_list, sg,
+                 status, s)
+{
+    if (is_ok(s)) {
+        fetch_and_add(&syslog.file_offset, bound(len));
+        if (syslog.file_offset >= SYSLOG_FILE_MAXSIZE) {
+            /* Do log rotation: rename the current log file by adding a ".1" extension to the file
+             * name, and continue logging on a new file. */
+            buffer archive_file = stack_allocate(sizeof(struct buffer));
```

Comment:
alloca_wrap_buffer cannot be used because we have to append a suffix to the file name, so we would be writing past the end of the original buffer contents.

Reply:
alloca_wrap_buffer does the same thing you're doing here...it's just eliminating a few lines of code. You're only overwriting the end of the buffer contents, not actually appending. So something like alloca_wrap_buffer(stack_allocate(file_path_len + 2), file_path_len + 2) and then just the runtime_memcpy and writing of last two chars should be equivalent.